rust/lance/src/io/writer/statistics.rs

```diff
@@ -0,0 +1,1082 @@
+    for idx in (0..data.len()).rev() {
+        let original = data[idx];
+        let (mut byte, mut overflow) = data[idx].overflowing_add(1);
+
+        // Until overflow: 0xFF -> 0x00
+        while !overflow {
+            data[idx] = byte;
+
+            if str::from_utf8(&data).is_ok() {
+                return Some(data);
+            }
+            (byte, overflow) = data[idx].overflowing_add(1);
+        }
+
+        data[idx] = original;
+    }
+
+    None
+}
+
+fn truncate_min_value(data: &[u8]) -> Vec<u8> {
+    match str::from_utf8(data) {
+        Ok(str_data) => truncate_utf8(str_data, BINARY_PREFIX_LENGTH),
+        Err(_) => truncate_binary(data, BINARY_PREFIX_LENGTH),
+    }
+    .unwrap_or_else(|| data.to_vec())
+}
+
+fn truncate_max_value(data: &[u8]) -> Vec<u8> {
+    match str::from_utf8(data) {
+        Ok(str_data) => truncate_utf8(str_data, BINARY_PREFIX_LENGTH).and_then(increment_utf8),
+        Err(_) => truncate_binary(data, BINARY_PREFIX_LENGTH).and_then(increment),
+    }
+    .unwrap_or_else(|| data.to_vec())
+}
+
+fn get_string_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
+    let mut min_value = vec![u8::MAX; BINARY_PREFIX_LENGTH];
+    let mut max_value = vec![u8::MIN; BINARY_PREFIX_LENGTH];
+    let mut null_count: i64 = 0;
+    let mut total_count: i64 = 0;
+
+    let arr = arrays
+        .iter()
+        .map(|x| x.as_any().downcast_ref::<StringArray>().unwrap())
+        .collect::<Vec<_>>();
+
+    for array in arr.iter() {
+        total_count += array.len() as i64;
+        array.iter().for_each(|value| {
+            if let Some(value) = value {
+                let value = value.as_bytes();
+                if let Some(Ordering::Greater) = value.partial_cmp(max_value.as_slice()) {
+                    max_value = value.to_vec();
+                }
+                if let Some(Ordering::Less) = value.partial_cmp(min_value.as_slice()) {
+                    min_value = value.to_vec();
+                }
+            }
+        });
+        null_count += array.null_count() as i64;
+    }
+
+    if total_count == 0 {
+        return StatisticsRow {
+            null_count: ScalarValue::Int64(Some(null_count)),
+            min_value: ScalarValue::Utf8(None),
+            max_value: ScalarValue::Utf8(None),
+        };
+    }
+
+    if max_value.len() > BINARY_PREFIX_LENGTH {
+        max_value = truncate_max_value(&max_value);
+    }
+    if min_value.len() > BINARY_PREFIX_LENGTH {
+        min_value = truncate_min_value(&min_value);
+    }
+    let min_value = str::from_utf8(&min_value).unwrap();
+    let max_value = str::from_utf8(&max_value).unwrap();
+
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(null_count)),
+        min_value: ScalarValue::Utf8(Some(min_value.to_string())),
+        max_value: ScalarValue::Utf8(Some(max_value.to_string())),
+    }
+}
+
+fn get_binary_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
```

Comment:
Much of the same feedback from the string stats applies here: don't create copies when writing to `min_value` and `max_value`, make sure we are returning null if all values are null, try to compare only the first N bytes instead of doing a memcmp on the whole buffer (this last one is even more important for binary, since these might be huge blobs).